### PR TITLE
TOOLS/PERF: Add wakeup option for UCP

### PR DESCRIPTION
**Pull request by Akshay-Venkatesh**
## What
As enabling UCP WAKEUP feature already exists, this PR adds poll/wait options for progressing transfers for UCP tests.

### CODE REVIEW

**Comment by Akshay-Venkatesh** [comment]
cc @yosefe @abellina

**Comment by Akshay-Venkatesh** [comment]
@yosefe Can you outline what needs to change in this PR and the new options say for Jenkins, multi-threaded testing, etc?

File: src/tools/perf/lib/ucp_tests.cc [hunk 1]
```diff
@@ -90,16 +90,32 @@ class ucp_perf_test_runner {
...
+        int count = 0;
+
+        if (UCX_PERF_WAIT_MODE_SLEEP == m_perf.params.wait_mode) {
+            while (ucp_worker_progress(worker)) {
+                count++;
+            }
+
+            if (!count) {
+                ucp_worker_wait(worker);
+            }
```
Comment (yosefe) [inline src/tools/perf/lib/ucp_tests.cc:103]: put this as a separate non-inline function with ucs_unlikely

**Review by yosefe — COMMENTED** [review]
pls add to gtest/ucp/test_ucp_perf.cc and perftest runs in test_jenkins.sh, but not to all tests (just 2 or 3) to avoid increasing testing time too much

**Comment by Akshay-Venkatesh** [comment]
> 
> 
> pls add to gtest/ucp/test_ucp_perf.cc and perftest runs in test_jenkins.sh, but not to all tests (just 2 or 3) to avoid increasing testing time too much

@yosefe noticing hangs in perftest with `ud`. Is UCP wakeup fully supported on `ud` alone?

**Comment by yosefe** [comment]
@Akshay-Venkatesh probably bug in UD. I'm ok to add a skip condition for wakeup tests over UD in this PR.

**Comment by Akshay-Venkatesh** [comment]
> 
> 
> @Akshay-Venkatesh probably bug in UD. I'm ok to add a skip condition for wakeup tests over UD in this PR.

I had to skip `shm` as well. I saw hangs there as well.

File: src/tools/perf/lib/ucp_tests.cc [hunk 2]
```diff
@@ -90,16 +90,36 @@ class ucp_perf_test_runner {
         }
     }
 
+    void make_blocking_progress() {
```
Comment (yosefe) [inline src/tools/perf/lib/ucp_tests.cc:93]: ALWAYS_INLINE blocking_progress

File: src/tools/perf/lib/ucp_tests.cc [hunk 3]
```diff
@@ -90,16 +90,36 @@ class ucp_perf_test_runner {
...
+        int count = 0;
+
+        while (ucp_worker_progress(m_perf.ucp.worker)) {
```
Comment (yosefe) [inline src/tools/perf/lib/ucp_tests.cc:96]: while (ucp_worker_progress(m_perf.ucp.worker) == 0)

File: src/tools/perf/lib/ucp_tests.cc [hunk 4]
```diff
@@ -90,16 +90,36 @@ class ucp_perf_test_runner {
...
+            count++;
+        }
+
+        if (!count) {
```
Comment (yosefe) [inline src/tools/perf/lib/ucp_tests.cc:100]: why needed?

File: src/tools/perf/lib/ucp_tests.cc [hunk 5]
```diff
@@ -90,16 +90,36 @@ class ucp_perf_test_runner {
...
+            ucp_worker_wait(m_perf.ucp.worker);
+        }
+    }
+
+    void UCS_F_ALWAYS_INLINE make_progress() {
```
Comment (yosefe) [inline src/tools/perf/lib/ucp_tests.cc:105]: just progress()

File: test/gtest/ucp/test_ucp_perf.cc [hunk 6]
```diff
@@ -57,115 +57,155 @@ const test_perf::test_spec test_ucp_perf::tests[] =
 {
   { "tag latency", "usec",
     UCX_PERF_API_UCP, UCX_PERF_CMD_TAG, UCX_PERF_TEST_TYPE_PINGPONG,
+    UCX_PERF_WAIT_MODE_LAST,
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_perf.cc:60]: UCX_PERF_WAIT_MODE_POLL
same for all others

File: test/gtest/ucp/test_ucp_perf.cc [hunk 7]
```diff
@@ -194,6 +234,13 @@ UCS_TEST_P(test_ucp_perf, envelope) {
          ++test_iter) {
         test_spec test = *test_iter;
 
+        if ((has_transport("ud_x") || has_transport("ud_v") ||
+             has_transport("ud") || has_transport("shm") ||
+             has_transport("self")) &&
+            (test.wait_mode == UCX_PERF_WAIT_MODE_SLEEP)) {
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_perf.cc:240]: pls add comment

File: test/gtest/uct/test_uct_perf.cc [hunk 8]
```diff
@@ -26,125 +26,146 @@ const test_perf::test_spec test_uct_perf::tests[] =
 {
   { "am short latency", "usec",
     UCX_PERF_API_UCT, UCX_PERF_CMD_AM, UCX_PERF_TEST_TYPE_PINGPONG,
+    UCX_PERF_WAIT_MODE_LAST,
```
Comment (yosefe) [inline test/gtest/uct/test_uct_perf.cc:29]: UCX_PERF_WAIT_MODE_POLL

File: test/gtest/ucp/test_ucp_perf.cc (hunk 6 above)
Comment (Akshay-Venkatesh) [reply test/gtest/ucp/test_ucp_perf.cc:60]: @yosefe 
MODE_POLL doesn't exist in libperf.h but MODE_PROGRESS exists. MODE_PROGRESS repeatedly calls progress. 
Are you suggesting to change name of MODE_LAST to MODE_POLL or are you suggesting to use MODE_PROGRESS?

Same for https://github.com/openucx/ucx/pull/6224#discussion_r573814460

Comment (yosefe) [reply test/gtest/ucp/test_ucp_perf.cc:60]: let's rename MODE_PROGRESS to MODE_POLL and use it
IMO, POLL is a clear and standard terminology

File: src/tools/perf/lib/ucp_tests.cc (hunk 3 above)
Comment (Akshay-Venkatesh) [reply src/tools/perf/lib/ucp_tests.cc:96]: @yosefe `make_blocking_progress` aims to drain the worker and call worker_wait if the worker is already drained (if `count == 0`). From the documentation of worker_wait, it looks like a requirement to drain worker before calling worker_wait. Wouldn't `while(ucp_worker_progress() == 0)` have the effect of polling?

File: src/tools/perf/lib/ucp_tests.cc (hunk 4 above)
Comment (Akshay-Venkatesh) [reply src/tools/perf/lib/ucp_tests.cc:100]: If `count` is zero, then worker is drained and it's safe to call worker_wait.

File: src/tools/perf/lib/ucp_tests.cc (hunk 3 above)
Comment (yosefe) [reply src/tools/perf/lib/ucp_tests.cc:96]: sorry, i meant != 0

File: src/tools/perf/lib/ucp_tests.cc (hunk 4 above)
Comment (yosefe) [reply src/tools/perf/lib/ucp_tests.cc:100]: if we get to this point, it means last call to progress returned 0

File: test/gtest/ucp/test_ucp_perf.cc (hunk 6 above)
Comment (Akshay-Venkatesh) [reply test/gtest/ucp/test_ucp_perf.cc:60]: Agreed. Will make the change now.

File: src/tools/perf/lib/ucp_tests.cc (hunk 3 above)
Comment (Akshay-Venkatesh) [reply src/tools/perf/lib/ucp_tests.cc:96]: @yosefe Looks like gtests hang if I use the following code:
```
while (ucp_worker_progress(m_perf.ucp.worker) != 0) {
}

ucp_worker_wait(m_perf.ucp.worker);
```
By forcing worker to drain completely and necessarily calling worker_wait, we are preventing either the requestor or responder from performing other transfers while being blocked in ucp_worker_wait. If the same thread is expected to perform the roles of both requestor and responder then draining worker and calling wait may result in a deadlock if the wakeup is also provided by the same thread, right?

Comment (yosefe) [reply src/tools/perf/lib/ucp_tests.cc:96]: how about
```
if (ucp_worker_progress() == 0) {
   ucp_worker_wait(m_perf.ucp.worker);
}
```

**Review by yosefe — APPROVED** [review]
pls squash

**Pull request merged**